core/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -56,18 +58,20 @@
     private final MetaStateService metaStateService;
     private final DanglingIndicesState danglingIndicesState;
     private final MetaDataIndexUpgradeService metaDataIndexUpgradeService;
+    private final PluginsService pluginsService;
 
     @Nullable
     private volatile MetaData previousMetaData;
 
     private volatile Set<Index> previouslyWrittenIndices = emptySet();
 
     @Inject
-    public GatewayMetaState(Settings settings, NodeEnvironment nodeEnv, MetaStateService metaStateService,
+    public GatewayMetaState(Settings settings, NodeEnvironment nodeEnv, PluginsService pluginsService, MetaStateService metaStateService,
```

Comment:
This is not the way to get the custom upgraders passed around (we should not be passing around PluginService). See other examples in Node: you should collect all the upgraders, and put them into a container that will then be injected (and eventually when GatewayMetaState is deguiced, we will just have it there as an arg, instead of via injection).